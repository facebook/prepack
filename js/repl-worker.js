self.importScripts('prepack.min.js');

onmessage = function(e) {
  var buffer = [];

  function errorHandler(error) {
    // Syntax errors contain their location at the end, remove that
    if (error.errorCode === 'PP1004') {
      let msg = error.message;
      error.message = msg.substring(0, msg.lastIndexOf('('));
    }
    buffer.push({
      severity: error.severity,
      location: error.location,
      message: error.message,
      errorCode: error.errorCode,
    });
    return 'Recover';
  }

  try {
    var sources = [{ filePath: 'dummy', fileContents: e.data }];
    var result = Prepack.prepackSources(sources, {
      compatibility: 'browser',
      filename: 'repl',
      timeout: 1000,
      serialize: true,
      errorHandler,
    });
    if (result && !buffer.length) {
      postMessage({ type: 'success', data: result.code });
    } else {
      // A well-defined error occurred.
      postMessage({ type: 'error', data: buffer });
    }
  } catch (err) {
    buffer.push({
      message: err.message || 'An unknown error occurred'
    });
    postMessage({ type: 'error', data: buffer });
  }
};
