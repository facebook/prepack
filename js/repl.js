function createEditor(elem) {
  var editor = ace.edit(elem);
  var session = editor.getSession();

  editor.setTheme('ace/theme/tomorrow');
  editor.setShowPrintMargin(false);
  editor.commands.removeCommands(['gotoline', 'find']);

  session.setMode('ace/mode/javascript');
  session.setUseSoftTabs(true);
  session.setTabSize(2);
  session.setUseWorker(false);

  editor.setOption('scrollPastEnd', 0.33);

  elem.style.lineHeight = '18px';

  return editor;
}

var demos = [];
/**generate select */
function generateSelect(obj, dom) {
  var tmpName;
  // var keys = ['<option value='+-1+'>select demo</option>'];
  var keys = [];
  demos = [];
  for (var name in obj) {
    if (!tmpName) {
      tmpName = name;
    }
    demos.push(name);
    keys.push('<option value=' + name + '>' + name + '</option>');
  }
  dom.innerHTML = keys.join('');
  return tmpName;
}

var worker;
var debounce;

var errorOutput = document.querySelector('.output .error');
var replOutput = document.querySelector('.output .repl');

var isEmpty = /^\s*$/;

function terminateWorker() {
  if (worker) {
    worker.terminate();
    worker = null;
  }
}

function processError(errorOutput, error) {
  let errorWikiLink = document.createElement('a');
  errorWikiLink.href = 'https://github.com/facebook/prepack/wiki/' + encodeURIComponent(error.errorCode);
  errorWikiLink.text = error.errorCode;
  errorWikiLink.setAttribute('target', '_blank');
  //console.log(JSON.stringify(error));
  // TODO: syntax errors need their location stripped
  if (error.location) {
    let errorLineLink = document.createElement('a');
    let lineNumber = error.location.start ? error.location.start.line : error.location.line;
    let colNumber = error.location.start ? error.location.start.column : error.location.column;
    colNumber++;
    let lineText = lineNumber + ':' + colNumber;
    errorLineLink.href = '';
    errorLineLink.onclick = function() {
      input.gotoLine(lineNumber);
      return false;
    };
    errorLineLink.text = lineText;
    errorLineLink.style.color = 'red';
    errorOutput.appendChild(errorWikiLink);
    errorOutput.appendChild(document.createTextNode(' ('));
    errorOutput.appendChild(errorLineLink);
    errorOutput.appendChild(document.createTextNode('):  ' + error.message + '\n'));
  } else if (!error.code) {
    errorOutput.appendChild(document.createTextNode(error.message + '\n'));
  } else {
    errorOutput.appendChild(errorWikiLink);
    errorOutput.appendChild(document.createTextNode(': ' + error.message + '\n'));
  }
}

function compile() {
  clearTimeout(debounce);
  terminateWorker();

  errorOutput.innerHTML = '';
  errorOutput.style.display = 'none';
  replOutput.style.display = 'block';

  output.setValue('// Compiling...', -1);

  debounce = setTimeout(function() {
    worker = new Worker('js/repl-worker.js');
    worker.onmessage = function(e) {
      // turn off compiling

      var result = e.data;
      if (result.type === 'success') {
        var code = result.data;
        if (isEmpty.test(code) && !isEmpty.test(input.getValue())) {
          code =
            '// Your code was all dead code and thus eliminated.\n' + '// Try storing a property on the global object.';
        }
        output.setValue(code, -1);
      } else if (result.type === 'error') {
        let errors = result.data;
        if (typeof errors === 'string') {
          errorOutput.style.display = 'block';
          replOutput.style.display = 'none';
          errorOutput.textContent = errors;
        } else {
          errorOutput.style.display = 'block';
          replOutput.style.display = 'none';
          for (var i in errors) {
            processError(errorOutput, errors[i]);
          }
        }
      }
      terminateWorker();
    };
    worker.postMessage(input.getValue());
  }, 500);
}

var output = createEditor(replOutput);
output.setReadOnly(true);
output.setHighlightActiveLine(false);
output.setHighlightGutterLine(false);

var input = createEditor(document.querySelector('.input .repl'));
input.on('change', compile);

/**record **/

var selectRecord = document.querySelector('select.select-record');
var selectInput = document.querySelector('#recordName');
var saveButton = document.querySelector('#saveBtn');
var deleteButton = document.querySelector('#deleteBtn');
var storage = window.localStorage;

var selectCache = getCache();
var defaultName, defaultVal;
defaultName = generateSelect(selectCache, selectRecord);
selectInput.value = defaultName || '';
defaultVal = defaultName ? selectCache[defaultName] : '';
input.setValue(defaultVal);
compile();

function changeSelect(val) {
  if (!val.value) return;
  selectInput.value = val.value;
  var localCache = getCache();
  var code = localCache[val.value] || '';
  input.setValue(code);
  compile();
}
changeSelect(defaultVal);

var demoSelector = new Select({
  el: selectRecord,
  className: 'select-theme-dark',
});
demoSelector.on('change', changeSelect);

function getCache() {
  return JSON.parse(storage.getItem('prepackDemos') || '{}');
}

function setCache(data) {
  storage.setItem('prepackDemos', JSON.stringify(data || {}));
}

function addDefaultExamples() {
  var cache = getCache();
  var code, name;
  name = 'EliminationOfAbstractionTax';
  code = [
    '(function () {',
    '  var self = this;',
    '    ["A", "B", 42].forEach(function(x) {',
    '    var name = "_" + x.toString()[0].toLowerCase();',
    '    var y = parseInt(x);',
    '    self[name] = y ? y : x;',
    '  });',
    '})();',
  ].join('\n');
  cache[name] = code;

  name = 'EnvironmentInteractionsAndBranching';
  code = [
    '(function(){',
    '  function fib(x) { return x <= 1 ? x : fib(x - 1) + fib(x - 2); }',
    '  let x = Date.now();',
    '  if (x === 0) x = fib(10);',
    '  global.result = x;',
    '})();',
  ].join('\n');
  cache[name] = code;

  name = 'Fibonacci';
  code = [
    '(function () {',
    '  function fibonacci(x) {',
    '    return x <= 1 ? x : fibonacci(x - 1) + fibonacci(x - 2);',
    '  }',
    '  global.x = fibonacci(10);',
    '})();',
  ].join('\n');
  cache[name] = code;

  name = 'HelloWorld';
  code = [
    '(function () {',
    '  function hello() { return "hello"; }',
    '  function world() { return "world"; }',
    '  global.s = hello() + " " + world();',
    '})();',
  ].join('\n');
  cache[name] = code;

  name = 'ModuleInitialization';
  code = [
    '(function () {',
    '  let moduleTable = {};',
    '  function define(id, f) { moduleTable[id] = f; }',
    '  function require(id) {',
    '    let x = moduleTable[id];',
    '    return x instanceof Function ? (moduleTable[id] = x()) : x;',
    '  }',
    '  global.require = require;',
    '  define("one", function() { return 1; });',
    '  define("two", function() { return require("one") + require("one"); });',
    '  define("three", function() { return require("two") + require("one"); });',
    '  define("four", function() { return require("three") + require("one"); });',
    '})();',
    'three = require("three");'
  ].join('\n');
  cache[name] = code;

  generateSelect(cache, selectRecord);
  setCache(cache);
  setTimeout(() => {
    demoSelector.change('Fibonacci');
  });
}

addDefaultExamples();

deleteButton.addEventListener('click', () => {
  var name = selectInput.value;
  if (name == null || name.replace(/\s+/, '') === '') return;
  if (demos.length === 0) {
    selectInput.value = '';
    return;
  }
  var cache = getCache();
  delete cache[name];
  input.setValue('');
  generateSelect(cache, selectRecord);
  setCache(cache);
  if (demos.length > 0) {
    selectInput.value = demos[0];
    setTimeout(() => {
      demoSelector.change(demos[0]);
    });
  } else {
    selectInput.value = '';
    input.setValue(defaultCode);
    compile();
  }
});

saveButton.addEventListener('click', () => {
  var name = selectInput.value;
  if (name == null || name.replace(/\s+/, '') === '') return;
  var code = input.getValue();
  var cache = getCache();
  cache[name] = code;
  generateSelect(cache, selectRecord);
  setCache(cache);
  setTimeout(() => {
    demoSelector.change(name);
  });
});
