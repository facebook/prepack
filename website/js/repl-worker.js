self.importScripts('prepack.min.js');

onmessage = function(e) {
  let buffer = [];

  function errorHandler(error) {
    // Syntax errors contain their location at the end, remove that
    if (error.errorCode === "PP1004") {
      let msg = error.message;
      error.message = msg.substring(0, msg.lastIndexOf("("));
    }
    buffer.push({
      severity: error.severity,
      location: error.location,
      message: error.message,
      errorCode: error.errorCode,
    });
    return "Recover";
  }

  try {
    let sources = [{ filePath: "dummy", fileContents: e.data.code }];
    let options = {
      compatibility: "browser",
      filename: "repl",
      timeout: 1000,
      serialize: true,
      heapGraphFormat: "VISJS",
      errorHandler,
    };
    for (let property in e.data.options) {
      if (e.data.options.hasOwnProperty(property)) {
        options[property] = e.data.options[property];
      }
    }
    let result = Prepack.prepackSources(sources, options);
    if (result && !buffer.length) {
      postMessage({ type: 'success', data: result.code, graph: result.heapGraph });
    } else {
      // A well-defined error occurred.
      postMessage({ type: 'error', data: buffer });
    }
  } catch (err) {
    buffer.push({
      message: err.message || 'An unknown error occurred'
    });
    postMessage({ type: 'error', data: buffer });
  }
};
