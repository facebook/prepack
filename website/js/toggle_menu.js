var toggle = document.getElementById("nav-toggle");
var menu = document.getElementById("nav-menu");

toggle.addEventListener("click", function() {
  if (this.className === "nav-toggle") {
    this.className = "nav-toggle is-active";
    menu.classList.add("is-active");
  } else {
    this.className = "nav-toggle";
    menu.classList.remove("is-active");
  }
});
