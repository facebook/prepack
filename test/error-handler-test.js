/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

/* @flow */

import { search, runTest } from "../scripts/test-error-handler.js";
const { it, expect } = global;

let tests = search(`${__dirname}/error-handler`, "test/error-handler");

for (let test of tests) {
  it(test.name, () => {
    runTest(test.name, test.file);
  })
}
