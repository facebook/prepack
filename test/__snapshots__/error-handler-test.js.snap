// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/error-handler/FinalObjectCannotBeMutated.js: test/error-handler/FinalObjectCannotBeMutated.js 1`] = `
Array [
  [Error: Mutating a final object, or an object with unknown properties, after some of those properties have already been used, is not supported.],
]
`;

exports[`test/error-handler/ModifiedObjectPropertyLimitation.js: test/error-handler/ModifiedObjectPropertyLimitation.js 1`] = `
Array [
  [Error: Program may terminate with exception],
  [Error: Internal slot $Prototype modified in a nested context. This is not yet supported.],
]
`;

exports[`test/error-handler/Set.js: test/error-handler/Set.js 1`] = `
Array [
  [Error: This version of JSC ignores the argument to Set, require the polyfill before doing this],
]
`;

exports[`test/error-handler/abstract-value-check-dup-name-string.js: test/error-handler/abstract-value-check-dup-name-string.js 1`] = `
Array [
  [Error: An abstract value with the same name exists],
]
`;

exports[`test/error-handler/bad-functions.js: test/error-handler/bad-functions.js 1`] = `
Array [
  [Error: Program may terminate with exception: Exception is not defined
  ReferenceError
    at additional1 (test/error-handler/bad-functions.js:7:27)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/binaryExpression.js: test/error-handler/binaryExpression.js 1`] = `
Array [
  [Error: might be an object with an unknown valueOf or toString or Symbol.toPrimitive method],
  [Error: might be an object with an unknown valueOf or toString or Symbol.toPrimitive method],
  [Error: might be an object with an unknown valueOf or toString or Symbol.toPrimitive method],
  [Error: might be an object with an unknown valueOf or toString or Symbol.toPrimitive method],
]
`;

exports[`test/error-handler/call.js: test/error-handler/call.js 1`] = `
Array [
  [Error: might not be a function],
  [Error: might not be a function],
  [Error: eval argument must be a known value],
]
`;

exports[`test/error-handler/call2.js: test/error-handler/call2.js 1`] = `
Array [
  [Error: Unfrozen object leaked before end of global code],
]
`;

exports[`test/error-handler/class.js: test/error-handler/class.js 1`] = `
Array [
  [Error: unknown super class],
  [Error: unknown super class prototype],
]
`;

exports[`test/error-handler/conditional-return.js: test/error-handler/conditional-return.js 1`] = `
Array [
  [Error: might be an object with an unknown valueOf or toString or Symbol.toPrimitive method],
]
`;

exports[`test/error-handler/forInStatement.js: test/error-handler/forInStatement.js 1`] = `
Array [
  [Error: for in loops over unknown objects are not yet supported],
]
`;

exports[`test/error-handler/forLoop1.js: test/error-handler/forLoop1.js 1`] = `
Array [
  [Error: break or continue with target cannot be guarded by abstract condition],
]
`;

exports[`test/error-handler/forOfStatement.js: test/error-handler/forOfStatement.js 1`] = `
Array [
  [Error: for of loops over unknown collections are not yet supported],
]
`;

exports[`test/error-handler/in1.js: test/error-handler/in1.js 1`] = `
Array [
  [Error: might not be an object, hence the in operator might throw a TypeError],
  [Error: might not be an object, hence the in operator might throw a TypeError],
]
`;

exports[`test/error-handler/in2.js: test/error-handler/in2.js 1`] = `
Array [
  [Error: might be an object that behaves badly for the in operator],
]
`;

exports[`test/error-handler/instanceof.js: test/error-handler/instanceof.js 1`] = `
Array [
  [Error: might be an object that behaves badly for the instanceof operator],
  [Error: might not be an object, hence the instanceof operator might throw a TypeError],
  [Error: might be an object that behaves badly for the instanceof operator],
]
`;

exports[`test/error-handler/member.js: test/error-handler/member.js 1`] = `
Array [
  [Error: member expression object abstract values true and ({})  is unknown],
]
`;

exports[`test/error-handler/member2.js: test/error-handler/member2.js 1`] = `
Array [
  [Error: member expression object abstract values true and ({})  is unknown],
]
`;

exports[`test/error-handler/object-assign.js: test/error-handler/object-assign.js 1`] = `
Array [
  [Error: This operation is not yet supported on abstract value ({foo:1}) ],
]
`;

exports[`test/error-handler/objectExpression.js: test/error-handler/objectExpression.js 1`] = `
Array [
  [Error: property key value is unknown],
  [Error: property key value is unknown],
]
`;

exports[`test/error-handler/objectExpression2.js: test/error-handler/objectExpression2.js 1`] = `
Array [
  [Error: property key value is unknown],
]
`;

exports[`test/error-handler/objectpattern.js: test/error-handler/objectpattern.js 1`] = `
Array [
  [Error: unknown computed property name],
]
`;

exports[`test/error-handler/require_throws2.js: test/error-handler/require_throws2.js 1`] = `
Array [
  [Error: Module import may fail with an exception],
]
`;

exports[`test/error-handler/syntaxError.js: test/error-handler/syntaxError.js 1`] = `
Array [
  [Error: Syntax error: Unexpected token, expected ( (4:3)],
]
`;

exports[`test/error-handler/testErrorHandlerCalled.js: test/error-handler/testErrorHandlerCalled.js 1`] = `
Array [
  [Error: This operation is not yet supported on abstract value  ],
]
`;

exports[`test/error-handler/try-and-access-abstract-property.js: test/error-handler/try-and-access-abstract-property.js 1`] = `
Array [
  [Error: Possible throw inside try/catch is not yet supported],
  [Error: Possible throw inside try/catch is not yet supported],
]
`;

exports[`test/error-handler/try-and-call-abstract-function.js: test/error-handler/try-and-call-abstract-function.js 1`] = `
Array [
  [Error: Possible throw inside try/catch is not yet supported],
]
`;

exports[`test/error-handler/unaryExpression.js: test/error-handler/unaryExpression.js 1`] = `
Array [
  [Error: might be a symbol or an object with an unknown valueOf or toString or Symbol.toPrimitive method],
  [Error: might be a symbol or an object with an unknown valueOf or toString or Symbol.toPrimitive method],
  [Error: might be a symbol or an object with an unknown valueOf or toString or Symbol.toPrimitive method],
]
`;

exports[`test/error-handler/updateExpression.js: test/error-handler/updateExpression.js 1`] = `
Array [
  [Error: might be a symbol or an object with an unknown valueOf or toString or Symbol.toPrimitive method],
  [Error: might be a symbol or an object with an unknown valueOf or toString or Symbol.toPrimitive method],
]
`;

exports[`test/error-handler/with.js: test/error-handler/with.js 1`] = `
Array [
  [Error: with object must be a known value],
]
`;

exports[`test/error-handler/with2.js: test/error-handler/with2.js 1`] = `
Array [
  [Error: with object must be a known value],
  [Error: This operation is not yet supported on (some value) at z],
]
`;

exports[`test/error-handler/write-forin-conflict.js: test/error-handler/write-forin-conflict.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-in-conflict.js: test/error-handler/write-in-conflict.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-reflect-conflict.js: test/error-handler/write-reflect-conflict.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-write-conflict.js: test/error-handler/write-write-conflict.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-write-conflict2.js: test/error-handler/write-write-conflict2.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-write-conflict3.js: test/error-handler/write-write-conflict3.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-write-conflict4.js: test/error-handler/write-write-conflict4.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;

exports[`test/error-handler/write-write-unknown-prop-conflict.js: test/error-handler/write-write-unknown-prop-conflict.js 1`] = `
Array [
  [Error: Property access conflicts with write in optimized function (unknown function)],
  [Error: Property access conflicts with write in optimized function (unknown function)],
]
`;
